test db9

```
2024-08-09 11:43:03.775 ERROR (Recorder) [homeassistant.helpers.recorder] Error executing query
Traceback (most recent call last):
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
    self.dialect.do_execute(
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/default.py", line 924, in do_execute
    cursor.execute(statement, parameters)
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/MySQLdb/cursors.py", line 179, in execute
    res = self._query(mogrified_query)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/MySQLdb/cursors.py", line 330, in _query
    db.query(q)
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/MySQLdb/connections.py", line 261, in query
    _mysql.connection.query(self, query)
MySQLdb.IntegrityError: (1452, 'Cannot add or update a child row: a foreign key constraint fails (`issue81989`.`#sql-alter-2f27-158`, CONSTRAINT `#sql-alter-2f27-158_ibfk_1` FOREIGN KEY (`old_state_id`) REFERENCES `states` (`state_id`))')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/bdraco/home-assistant/homeassistant/helpers/recorder.py", line 101, in session_scope
    yield session
  File "/Users/bdraco/home-assistant/homeassistant/components/recorder/migration.py", line 713, in _add_constraint
    connection.execute(add_constraint)
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1418, in execute
    return meth(
           ^^^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/sql/ddl.py", line 180, in _execute_on_connection
    return connection._execute_ddl(
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1529, in _execute_ddl
    ret = self._execute_context(
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1846, in _execute_context
    return self._exec_single_context(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1986, in _exec_single_context
    self._handle_dbapi_exception(
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 2353, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
    self.dialect.do_execute(
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/default.py", line 924, in do_execute
    cursor.execute(statement, parameters)
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/MySQLdb/cursors.py", line 179, in execute
    res = self._query(mogrified_query)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/MySQLdb/cursors.py", line 330, in _query
    db.query(q)
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/MySQLdb/connections.py", line 261, in query
    _mysql.connection.query(self, query)
sqlalchemy.exc.IntegrityError: (MySQLdb.IntegrityError) (1452, 'Cannot add or update a child row: a foreign key constraint fails (`issue81989`.`#sql-alter-2f27-158`, CONSTRAINT `#sql-alter-2f27-158_ibfk_1` FOREIGN KEY (`old_state_id`) REFERENCES `states` (`state_id`))')
[SQL: ALTER TABLE states ADD FOREIGN KEY(old_state_id) REFERENCES states (state_id)]
(Background on this error at: https://sqlalche.me/e/20/gkpj)
2024-08-09 11:43:03.780 ERROR (Recorder) [homeassistant.components.recorder.migration] Could not update foreign options in states table, will delete violations and try again
Traceback (most recent call last):
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
    self.dialect.do_execute(
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/default.py", line 924, in do_execute
    cursor.execute(statement, parameters)
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/MySQLdb/cursors.py", line 179, in execute
    res = self._query(mogrified_query)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/MySQLdb/cursors.py", line 330, in _query
    db.query(q)
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/MySQLdb/connections.py", line 261, in query
    _mysql.connection.query(self, query)
MySQLdb.IntegrityError: (1452, 'Cannot add or update a child row: a foreign key constraint fails (`issue81989`.`#sql-alter-2f27-158`, CONSTRAINT `#sql-alter-2f27-158_ibfk_1` FOREIGN KEY (`old_state_id`) REFERENCES `states` (`state_id`))')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/bdraco/home-assistant/homeassistant/components/recorder/migration.py", line 691, in _restore_foreign_key_constraints
    _add_constraint(session_maker, add_constraint, table)
  File "/Users/bdraco/home-assistant/homeassistant/components/recorder/migration.py", line 713, in _add_constraint
    connection.execute(add_constraint)
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1418, in execute
    return meth(
           ^^^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/sql/ddl.py", line 180, in _execute_on_connection
    return connection._execute_ddl(
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1529, in _execute_ddl
    ret = self._execute_context(
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1846, in _execute_context
    return self._exec_single_context(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1986, in _exec_single_context
    self._handle_dbapi_exception(
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 2353, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
    self.dialect.do_execute(
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/sqlalchemy/engine/default.py", line 924, in do_execute
    cursor.execute(statement, parameters)
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/MySQLdb/cursors.py", line 179, in execute
    res = self._query(mogrified_query)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/MySQLdb/cursors.py", line 330, in _query
    db.query(q)
  File "/Users/bdraco/home-assistant/venv/lib/python3.12/site-packages/MySQLdb/connections.py", line 261, in query
    _mysql.connection.query(self, query)
sqlalchemy.exc.IntegrityError: (MySQLdb.IntegrityError) (1452, 'Cannot add or update a child row: a foreign key constraint fails (`issue81989`.`#sql-alter-2f27-158`, CONSTRAINT `#sql-alter-2f27-158_ibfk_1` FOREIGN KEY (`old_state_id`) REFERENCES `states` (`state_id`))')
[SQL: ALTER TABLE states ADD FOREIGN KEY(old_state_id) REFERENCES states (state_id)]
(Background on this error at: https://sqlalche.me/e/20/gkpj)
```
```
| 345 | issue81989 | localhost | issue81989 | Query   | 9    | Sending data | DELETE FROM states WHERE (  states.old_state_id IS NOT NULL AND  NOT EXISTS    (SELECT 1      FROM  states      WHERE states.state_id = states.old_state_id)) | 9390.204  | 0     | 0         | 0.000    | 2085480     | 2478712         | 13164853      | 0         | 1657869  | DELETE FROM states WHERE (  states.old_state_id IS NOT NULL AND  NOT EXISTS    (SELECT 1      FROM  states      WHERE states.state_id = states.old_state_id)) | 0   |
```

--- a/homeassistant/components/recorder/migration.py
+++ b/homeassistant/components/recorder/migration.py
@@ -728,13 +728,13 @@ def _delete_foreign_key_violations(
     with session_scope(session=session_maker()) as session:
         session.execute(
             text(
-                f"DELETE FROM {table} "  # noqa: S608
+                f"DELETE FROM {table} t1"  # noqa: S608
                 "WHERE ("
-                f"  {table}.{column} IS NOT NULL AND"
+                f"  t1.{column} IS NOT NULL AND"
                 "  NOT EXISTS"
                 "    (SELECT 1 "
-                f"     FROM  {foreign_table} "
-                f"     WHERE {foreign_table}.{foreign_column} = {table}.{column}));"
+                f"     FROM  {foreign_table} t2"
+                f"     WHERE t2.{foreign_column} = t1.{column}));"
             )
         )
 

--- a/homeassistant/components/recorder/migration.py
+++ b/homeassistant/components/recorder/migration.py
@@ -729,7 +729,7 @@ def _delete_foreign_key_violations(
         session.execute(
             text(
                 f"DELETE FROM {table} t1"  # noqa: S608
-                "WHERE ("
+                " WHERE ("
                 f"  t1.{column} IS NOT NULL AND"
                 "  NOT EXISTS"
                 "    (SELECT 1 "

--- a/tests/components/recorder/test_migrate.py
+++ b/tests/components/recorder/test_migrate.py
@@ -1002,7 +1002,6 @@ def test_restore_foreign_key_constraints_with_integrity_error(
     assert len(connection.execute.mock_calls) == 2
     session.execute.assert_called_once()
     assert session.execute.mock_calls[0][1][0].text == (
-        "DELETE FROM events WHERE (  events.data_id IS NOT NULL AND  NOT EXISTS    "
-        "(SELECT 1      FROM  event_data      WHERE event_data.data_id = "
-        "events.data_id));"
+        "DELETE FROM events t1 WHERE (  t1.data_id IS NOT NULL AND  NOT EXISTS    "
+        "(SELECT 1      FROM  event_data t2     WHERE t2.data_id = t1.data_id));"
     )

